Instructor: edit instructors: reflect changes to not-yet-joined instructor's visibility elsewhere in the system #7537

HirdayGupta opened the pull request:
Fixes #7537 

After looking in all the wrong places, I seem to have finally found the error.

**The Problem**:

Those instructors who have not yet joined, their entities are updated by calling the `updateInstructorByEmail()` method in `InstructorsDb.java` as opposed to the `updateInstructorsByGoogleId()` method for instructors who have joined. 

However, in the `updateInstructorByEmail()` method, the `isDisplayedToStudents` property was not being updated, and was instead being carried forward from its previous state. This was the cause of the error.

**Outline of solution**:

Added a line that updates the `isDisplayedToStudents` property in `updateInstructorByEmail()`

damithc commented:
@HirdayGupta you can consider rebasing your branch instead of merging master to your branch. That way, you wont end up with so many 'merge commits' in your branch.

HirdayGupta commented:
> you can consider rebasing your branch instead of merging master to your branch. That way, you wont end up with so many 'merge commits' in your branch.
@damithc any suggestions on how I should go about doing this? I entered `git rebase master` with my branch checked out, and instead of the `merge commit` or "hiding" the `merge commits` it just rewinded to the HEAD, applied all the unmerged commits as fresh individual commits and put my commit on top.

damithc commented:
@HirdayGupta typically, we rebase _instead_ of merging master. So, my suggestion was for future PRs.

HirdayGupta commented:
>typically, we rebase instead of merging master. So, my suggestion was for future PRs.

@damithc Ahh, I see. Understood. Will look into rebasing and ask further question as dev help in the issue tracker. Thanks for the suggestion, was going to ask how other PRs do without the annoying `merge commits` anyway. 😄 

Meanwhile, this PR is **Ready for review**

leeyimin commented:
@HirdayGupta It will be good to add a test for this!

HirdayGupta commented:
@leeyimin Should I just be running expected HTML tests to see if the changes occured (before and after the changes)? Or should I be making a change and then `assertrue` that the change was made by calling the datastore?

leeyimin commented:
@HirdayGupta I believe you should at least update InstructorsDbTest and I would encourage you to add an additional UI test to ensure the bug you caught does not happen!

HirdayGupta commented:
@leeyimin Added a test for `isDisplayedToStudents` to `InstructorsDbTest`. 

However, I feel that this bug arose only because of how sparsely updating of instructors is tested. Currently, only updating the googleid and name of the instructor is tested. I believe we should also add tests that check the updating of each of the editable instructor parameters. Would you agree with this? And if yes, should I open a new issue for this or make the necessary changes in this PR itself (since it is a bug fix and prevention PR)

Also,

> I would encourage you to add an additional UI test to ensure the bug you caught does not happen!

I don't think I fully understood what you meant, so just to clarify, you are recommending a before edit and after edit HTML verification to ensure the changes are reflected?

leeyimin commented:
@HirdayGupta I agree with you on the first point. I would think it's fine to do it in this PR as well. 

Looking at the current InstructorCourseEditPageUiTest, I would think that HTML verification is not the best way to go about doing it. You could update the `testEditInstructorAction` and test editing on more fields and assert that the relevant UI fields contain the expected values after the edit. Alternatively, you could add a test case in InstructorCourseInstructorEditSaveActionTest. The first option would ensure that both the front end and back end systems are working properly to support an edit action but it would take more effort to do it too 😂

HirdayGupta commented:
**@leeyimin Ready for review!** 

Added comprehensive tests to InstructorCourseEditPageUiTest to test the editing of both joined and yet-to-join instructors and verify if the changes made match the edits. 

Please note that between https://github.com/TEAMMATES/teammates/pull/7559/commits/86898a2bca8c3224e99ded92f7198d1cf32c6349 and https://github.com/TEAMMATES/teammates/pull/7559/commits/b7af2d6d25aa8c1cee2ba660270ce9744b470f43, I changed my approach and moved all the `assertEquals` to a method in InstructorCourseEditPage as I would be reusing them across all the new test sections I added.

Additional note: Let me know if you would like to see smaller commits. I was unsure of when to commit each change that I was making. 😅

leeyimin commented on src/test/java/teammates/test/cases/storage/InstructorsDbTest.java:
```diff
@@ -314,12 +314,20 @@ public void testUpdateInstructorByGoogleId() throws Exception {
+        instructorToEdit.role = "new role";
+        instructorToEdit.isDisplayedToStudents = false;
+        instructorToEdit.displayedName = "New Displayed Name";
         instructorsDb.updateInstructorByGoogleId(instructorToEdit);
```
How about updating privileges as well?

HirdayGupta replied:
This is what a `privelegesAsText` string would look like:
```JAVA
{
  "courseLevel": {
    "canviewstudentinsection": true,
    "cansubmitsessioninsection": true,
    "canmodifysessioncommentinsection": true,
    "canmodifycourse": false,
    "canviewsessioninsection": true,
    "canmodifysession": true,
    "canmodifystudent": true,
    "canmodifyinstructor": true
  },
  "sectionLevel": {},
  "sessionLevel": {}
}
```

Creating such a large string in the method itself would not read very well in my opinion. How would you suggest changing this? (or is there another method to edit privileges that I am missing?O)

leeyimin replied:
An easy way to change privileges would be using the InstructorPrivileges constructor at line 104, there are a few defined settings.

HirdayGupta commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +184,23 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
+        assertEquals(newName, getInstructorName(instrNum));
+        if (!newEmail.isEmpty()) {
+            assertEquals(newEmail, getInstructorEmail(instrNum));
+        }
```
The problem with that arises from the test case where we are making changes to a yet-to-join instructor. Here, the email field is readonly and so if I try to edit it, the test throws an error. Hence I used this and passed a blank string when editing the email of a yet-to-join instructor.

HirdayGupta replied:
Wait, ignore that last comment, my mistake. I thought this comment was about the `if (getEmailField(instrNum).getAttribute("readonly") == null)` in the same file :p) I will assert the same unchanged value in the case of the yet-to-join instructor.

HirdayGupta replied:
fixed in https://github.com/TEAMMATES/teammates/pull/7559/commits/6bc9633b25bfc87bb94b016ba8227f0412d912da

HirdayGupta commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +184,23 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
+            assertTrue(getInstructorDisplayedToStudents(instrNum));
+            assertEquals(newDisplayName, getInstructorDisplayName(instrNum));
+        } else {
+            assertTrue(!getInstructorDisplayedToStudents(instrNum));
```
For some strange reason, `assertFalse` does not resolve to a type here 😂

leeyimin replied:
You can do a static import, the same done for other assert methods.
```
import static org.testng.AssertJUnit.assertFalse;
```

HirdayGupta commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -90,11 +90,17 @@ public void addNewInstructor(String name, String email) {
     }
 
-    public void editInstructor(int instrNum, String name, String email, String role) {
+    public void editInstructor(int instrNum, String name, String email,
```
Fixed in https://github.com/TEAMMATES/teammates/pull/7559/commits/cf41d1694dc04925b4bf436fb698203a993d7f7a and https://github.com/TEAMMATES/teammates/pull/7559/commits/b120a0141e98a7ebd609dca9e3cee1333aff8447

HirdayGupta commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -90,11 +90,17 @@ public void addNewInstructor(String name, String email) {
+        if (getEmailField(instrNum).getAttribute("readonly") == null) {
+            editInstructorEmail(instrNum, email);
+        }
+        if (editInstructorDisplayedToStudents(instrNum, isDisplayedToStudents)) {
```
Fixed in https://github.com/TEAMMATES/teammates/pull/7559/commits/fbb08257197813ea23abb7bc1d9c90004badf3e4

HirdayGupta commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +184,23 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
         assertTrue(isNotEditable);
     }
 
+    public void verifyInstructorChangesAfterEdit(int instrNum, String newName, String newEmail,
```
Renamed in https://github.com/TEAMMATES/teammates/pull/7559/commits/f2fb267d58bef77075d037a9ede2db1368a1b08a and added documentation in https://github.com/TEAMMATES/teammates/pull/7559/commits/8774251e0c0e4f68163cbbb70aa61b371e9fc51b

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -90,11 +90,22 @@ public void addNewInstructor(String name, String email) {
     }
 
-    public void editInstructor(int instrNum, String name, String email, String role) {
+    /*
```
Use the java doc comment, `/**`.

leeyimin replied:
Should start with a general statement of what the method does.

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +189,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
         assertTrue(isNotEditable);
     }
 
+    /*
```
Same here :)

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +189,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
     }
 
+    /*
+     * newDisplayName is verified depending on newIsDisplayedToStudents.
```
You can use `{@code newDisplayName}` and `{@code newIsDisplayedToStudents}` to refer to the parameters.

HirdayGupta commented:
@leeyimin All changes made. Ready for review! 😃

leeyimin commented on src/test/java/teammates/test/cases/storage/InstructorsDbTest.java:
```diff
@@ -328,6 +330,7 @@ public void testUpdateInstructorByGoogleId() throws Exception {
         assertEquals("new role", instructorUpdated.role);
         assertFalse(instructorUpdated.isDisplayedToStudents);
         assertEquals("New Displayed Name", instructorUpdated.displayedName);
+        assertTrue(instructorUpdated.hasObserverPrivileges());
```
In addition to that, you might want to do a `assertFalse(instructorUpdated.has<original-role>Privileges());` since a more privileged role is likely to return true for `hasObserverPrivileges`.

HirdayGupta replied:
@leeyimin I tried that out and it doesn't return true for more privileged roles but I guess I'll add this in any way to protect from any future changes.

leeyimin commented on src/test/java/teammates/test/cases/storage/InstructorsDbTest.java:
```diff
@@ -397,6 +402,7 @@ public void testUpdateInstructorByEmail() throws Exception {
         assertEquals("new role", instructorUpdated.role);
         assertFalse(instructorUpdated.isDisplayedToStudents);
         assertEquals("New Displayed Name", instructorUpdated.displayedName);
+        assertTrue(instructorUpdated.hasObserverPrivileges());
```
Same here.

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +191,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
     }
 
+    /**
+     * {@code newDisplayName} is verified depending on {@code newIsDisplayedToStudents}.
```
Should start with a general statement of what the method does.

HirdayGupta replied:
@leeyimin Have deleted the existing first line of the Javadoc comment (as it is basically repeated in the next two lines) and replaced it with the general statement in 0198869

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -411,6 +460,12 @@ public WebElement getDisplayedToStudentCheckBox(int instrNum) {
+    public WebElement getDisplayNameField(int instrNum) {
+        return browser.driver.findElement(By.cssSelector("#instructorTable" + instrNum + " input[name='"
+                + Const.ParamsNames.INSTRUCTOR_DISPLAY_NAME
+                + "']"));
```
Extract the string or the expression `By.cssSelector(...)` as a variable for better readability.

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -419,6 +474,22 @@ public String getInstructorEmail(int instrNum) {
+    public boolean getInstructorDisplayedToStudents(int instrNum) {
+        return browser.driver.findElement(By.cssSelector("#instructorTable" + instrNum + " input[name='"
+                + Const.ParamsNames.INSTRUCTOR_IS_DISPLAYED_TO_STUDENT
+                + "']")).isSelected();
```
Same here.

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -419,6 +474,22 @@ public String getInstructorEmail(int instrNum) {
+    public String getInstructorDisplayName(int instrNum) {
+        return browser.driver.findElement(By.cssSelector("#instructorTable" + instrNum + " input[name='"
+                + Const.ParamsNames.INSTRUCTOR_DISPLAY_NAME
+                + "']")).getAttribute("value");
```
Same here.

HirdayGupta commented:
@leeyimin All changes made. Ready for review!

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +191,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
 
+    /**
+     * Verifies that the fields contain the updated values.
+     * If {@code newIsDisplayedToStudents} is true, the field's value is checked against {@code newDisplayName}.
```
`the display name field is checked against {@code newDisplayName}` instead? What `the field` is referring to is ambiguous.

HirdayGupta commented:
@leeyimin Ready for review

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +191,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
     }
 
+    /**
+     * Verifies that the Instructor Details fields for the given {@code instrNum} contain the updated values.
```
`Instructor Details` do not have to be capitalized here.

HirdayGupta replied:
@leeyimin oops my bad. Sorry for the sloppiness. Fixed.

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +191,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
+     * Otherwise, it is verified that the display name field's placeholder is shown as expected.
+     */
+    public void verifyInstructorDetails(int instrNum, String newName, String newEmail,
+            Boolean newIsDisplayedToStudents, String newDisplayName, String newRole) {
```
Might want to use `boolean` instead.

HirdayGupta replied:
@leeyimin A boolean instead of the `void` return type? Not too sure what you meant here. 😅

leeyimin replied:
`boolean newIsDisplayedToStudents`

leeyimin replied:
This is so that the next comment will make more sense haha.

leeyimin commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -162,6 +191,26 @@ public void verifyInstructorEditFormDisabled(int instrNum) {
+        assertEquals(newName, getInstructorName(instrNum));
+        assertEquals(newEmail, getInstructorEmail(instrNum));
+        if (newIsDisplayedToStudents) {
+            assertTrue(getInstructorDisplayedToStudents(instrNum));
```
This line can be extracted out and explicitly asserted against `newIsDisplayedToStudents`.

leeyimin commented on src/test/java/teammates/test/cases/storage/InstructorsDbTest.java:
```diff
@@ -314,12 +314,24 @@ public void testUpdateInstructorByGoogleId() throws Exception {
+        assertFalse(instructorUpdated.isDisplayedToStudents);
+        assertEquals("New Displayed Name", instructorUpdated.displayedName);
+        assertTrue(instructorUpdated.hasObserverPrivileges());
+        assertFalse(instructorUpdated.hasManagerPrivileges());
```
Add some documentation explaining why this is done.

HirdayGupta replied:
Would this be okay:
```JAVA
assertTrue(instructorUpdated.hasObserverPrivileges());
// Verifying less privileged 'Observer' role did not return false positive in case old 'Manager' role is unchanged.
assertFalse(instructorUpdated.hasManagerPrivileges());
```

leeyimin commented on src/test/java/teammates/test/cases/storage/InstructorsDbTest.java:
```diff
@@ -375,12 +387,24 @@ public void testUpdateInstructorByEmail() throws Exception {
+        assertFalse(instructorUpdated.isDisplayedToStudents);
+        assertEquals("New Displayed Name", instructorUpdated.displayedName);
+        assertTrue(instructorUpdated.hasObserverPrivileges());
+        assertFalse(instructorUpdated.hasCoownerPrivileges());
```
Here too!

HirdayGupta commented:
@leeyimin Ready for review
